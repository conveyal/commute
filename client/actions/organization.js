import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.analyses = []
  newOrganization.analysesById = {}
  newOrganization.sites = []
  newOrganization.sitesById = {}
  newOrganization.groups = []
  newOrganization.groupsById = {}
  return [
    addLocally(newOrganization),
    push(`/organizations/${newOrganization.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete organization')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/organization/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteOrganization = (id) => [deleteLocally(id), push('/')]

export const updateOrganization = (organization) => [
  setLocally(organization),
  push(`/organizations/${organization.id}`)
]
/* const updateOnServer = (organization) =>
  serverAction({
    url: `/api/organization/${organization.id}`,
    options: {
      body: JSON.stringify(organization),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set organization')
