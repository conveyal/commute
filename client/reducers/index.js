import {handleActions} from 'redux-actions'

import * as organization from './organization'
import * as user from './user'

export default {
  organization: handleActions(organization.reducers, organization.initialState),
  user: handleActions(user.reducers, user.initialState)
}
