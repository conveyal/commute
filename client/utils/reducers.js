export function arrayToObj (a) {
  const obj = {}
  for (let i = 0; i < a.length; i++) obj[a[i].id] = a[i]
  return obj
}

/**
 * Round a number to a fixed amount of decimal places
 *
 * @param  {Number} n      The number to round
 * @param  {Number} places The number of places to round to (default=2)
 * @return {Number}        The rounded number
 */
export function fixedRound (n, places) {
  places = places || 2
  const multiplier = Math.pow(10, places)
  return Math.round(n * multiplier) / multiplier
}
