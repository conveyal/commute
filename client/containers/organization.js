import {connect} from 'react-redux'

import {deleteAnalysis} from '../actions/analysis'
import {deleteGroup} from '../actions/group'
import {deleteOrganization} from '../actions/organization'
import {deleteSite} from '../actions/site'
import Organization from '../components/organization'

function mapStateToProps (state, props) {
  const {organization} = state
  const {params} = props
  const currentOrganization = organization.organizationsById[params.organizationId]
  return {organization: currentOrganization}
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (id, organizationId) => dispatch(deleteAnalysis(id, organizationId)),
    deleteGroup: (id, organizationId) => dispatch(deleteGroup(id, organizationId)),
    deleteOrganization: (id) => dispatch(deleteOrganization(id)),
    deleteSite: (id, organizationId) => dispatch(deleteSite(id, organizationId))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Organization)
