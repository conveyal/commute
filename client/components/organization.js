import React, {Component, PropTypes} from 'react'
import {Button, ButtonGroup, Col, Grid, Row} from 'react-bootstrap'
import {BootstrapTable, TableHeaderColumn} from 'react-bootstrap-table'
import {Link} from 'react-router'

import Icon from './icon'
import {messages} from '../utils/env'
import {arrayCountRenderer} from '../utils/table'
import {actUponConfirmation} from '../utils/ui'

export default class Organization extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,
    deleteGroup: PropTypes.func,
    deleteOrganization: PropTypes.func,
    deleteSite: PropTypes.func,

    // props
    organization: PropTypes.object
  }

  _analysisGroupNameRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    const group = this.props.organization.groupsById[row.siteId]
    return <Link to={`/organizations/${organizationId}/groups/${group.id}`}>{group.name}</Link>
  }

  _analysisNameRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    return <Link to={`/organizations/${organizationId}/analysis/${row.id}`}>{row.name}</Link>
  }

  _analysisSiteNameRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    const site = this.props.organization.sitesById[row.siteId]
    return <Link to={`/organizations/${organizationId}/groups/${site.id}`}>{site.name}</Link>
  }

  _analysisToolsRenderer = (cell, row) => {
    const doDelete = () => {
      this.props.deleteAnalysis(row.id, this.props.organization.id)
    }
    const onClick = () => actUponConfirmation(messages.analysis.deleteConfirmation, doDelete)
    return <Button bsStyle='danger' onClick={onClick}>Delete</Button>
  }

  _handleDelete = () => {
    const doDelete = () => this.props.deleteOrganization(this.props.organization.id)
    actUponConfirmation(messages.organization.deleteConfirmation, doDelete)
  }

  _groupNameRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    return <Link to={`/organizations/${organizationId}/groups/${row.id}/`}>{row.name}</Link>
  }

  _groupToolsRenderer = (cell, row) => {
    const doDelete = () => this.props.deleteGroup(row.id, this.props.organization.id)
    const onClick = () => actUponConfirmation(messages.group.deleteConfirmation, doDelete)
    return <Button bsStyle='danger' onClick={onClick}>Delete</Button>
  }

  _siteNameRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    return <Link to={`/organizations/${organizationId}/sites/${row.id}/`}>{row.name}</Link>
  }

  _siteToolsRenderer = (cell, row) => {
    const organizationId = this.props.organization.id
    const doDelete = () => this.props.deleteSite(row.id, organizationId)
    const onClick = () => actUponConfirmation(messages.site.deleteConfirmation, doDelete)
    return <div>
      <Button bsStyle='warning'>
        <Link to={`/organizations/${organizationId}/sites/${row.id}/edit`}>Edit</Link>
      </Button>
      <Button bsStyle='danger' onClick={onClick}>Delete</Button>
    </div>
  }

  render () {
    const {analyses, id: _id, groups, name, sites} = this.props.organization

    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h2><Icon type='shield' />{name}</h2>
            <ButtonGroup>
              <Button bsStyle='warning'>
                <Link to={`/organizations/${_id}/edit`}>Edit</Link>
              </Button>
              <Button bsStyle='danger' onClick={this._handleDelete}>Delete</Button>
            </ButtonGroup>
            <p>Below are this organization's sites, commuter groups and analyses.</p>
            <h3>Sites
              <Button className='pull-right'>
                <Link to={`/organizations/${_id}/sites/create`}>Create a new site <Icon type='building' /></Link>
              </Button>
            </h3>
            <p>A site is a location of a building or new address that you want to use as the centerpoint of your commutes.</p>
            <BootstrapTable data={sites}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._siteNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='address'>Address</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._siteToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
            <h3>Commuter Groups
              <Button className='pull-right'>
                <Link to={`/organizations/${_id}/groups/create`}>Create a new group <Icon type='users' /></Link>
              </Button>
            </h3>
            <p>A commuter group is a list of commuters that can commute to a particular site.</p>
            <BootstrapTable data={groups}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._groupNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataField='commuters' dataFormat={arrayCountRenderer}>Commuters</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._groupToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
            <h3>Analyses
              <Button className='pull-right'>
                <Link to={`/organizations/${_id}/analysis/create`}>Create a new analysis <Icon type='bar-chart' /></Link>
              </Button>
            </h3>
            <p>An analysis calculates commuting statistics for a pairing of a commuter group and site.</p>
            <BootstrapTable data={analyses}>
              <TableHeaderColumn dataField='id' isKey hidden />
              <TableHeaderColumn dataFormat={this._analysisNameRenderer}>Name</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisSiteNameRenderer}>Site</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisGroupNameRenderer}>Group</TableHeaderColumn>
              <TableHeaderColumn dataFormat={this._analysisToolsRenderer}>Tools</TableHeaderColumn>
            </BootstrapTable>
          </Col>
        </Row>
      </Grid>
    )
  }
}
