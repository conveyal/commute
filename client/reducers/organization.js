import update from 'react-addons-update'

import {settings} from '../utils/env'
import {arrayToObj, fixedRound} from '../utils/reducers'

const METRICS = Object.keys(settings.metrics)

export const reducers = {

  /**
   * Organization Stuff
   */

  'add organization' (state, action) {
    return addEntities(state, 'organizations', [action.payload])
  },
  'delete organization' (state, action) {
    return deleteEntity(state, 'organizations', action.payload.id)
  },
  'set organization' (state, action) {
    return setOrganization(state, action.payload)
  },
  'set organizations' (state, action) {
    const organizations = [...action.payload]
    return {
      organizations,
      organizationsById: arrayToObj(action.payload)
    }
  },

  /**
   * Analysis Stuff
   */

  'add analysis' (state, action) {
    const {organizationId, analysis} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'analyses', [analysis])
    return setOrganization(state, updatedOrganization)
  },
  'delete analysis' (state, action) {
    const {organizationId, analysisId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'analyses', analysisId)
    return setOrganization(state, updatedOrganization)
  },
  'receive mock calculated trips' (state, action) {
    const {analysisId, organizationId, trips} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedAnalysis = affectedOrganization.analysesById[analysisId]
    let updatedAnalysis = update(affectedAnalysis, {
      trips: { $set: trips }
    })

    // calculate array of values for each metric
    const vals = {}
    const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
    modes.forEach((mode) => {
      vals[mode] = {
        cost: [],
        distance: [],
        time: []
      }
    })
    trips.forEach((trip) =>
      modes.forEach((mode) =>
        METRICS.forEach((metric) => {
          vals[mode][metric].push(trip[mode][metric])
        })
      )
    )

    // sort arrays
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].sort()
      })
    )

    updatedAnalysis = update(updatedAnalysis, {
      tripVals: { $set: vals }
    })

    // calculate analysis summary metrics
    // Potential Savings is an overall average of the difference in cost between
    // the most likely trip and the driving trip
    const len = trips.length
    let totalCostDifference = 0
    let totalTravelTime = 0
    let totalDistance = 0
    trips.forEach((trip) => {
      totalCostDifference += trip.car.cost - trip.mostLikely.cost
      totalTravelTime += trip.mostLikely.time
      totalDistance += trip.mostLikely.distance
    })

    updatedAnalysis = update(updatedAnalysis, {
      summary: {
        $set: {
          avgTravelTime: fixedRound(totalTravelTime / len),
          avgDistance: fixedRound(totalDistance / len),
          savingsPerTrip: fixedRound(totalCostDifference / len),
          savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
          savingsTotalPerDay: fixedRound(totalCostDifference),
          savingsTotalPerYear: fixedRound(totalCostDifference * 260)
        }
      }
    })

    const updatedOrganization = setEntity(affectedOrganization, 'analyses', updatedAnalysis)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Commuter Stuff
   */

  'add commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', [commuter])
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete commuter' (state, action) {
    const {commuterId, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = deleteEntity(affectedGroup, 'commuters', commuterId)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'update commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = setEntity(affectedGroup, 'commuters', commuter)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Group Stuff
   */

  'add group' (state, action) {
    const {organizationId, group} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    group.commutersById = arrayToObj(group.commuters)
    const updatedOrganization = addEntities(affectedOrganization, 'groups', [group])
    return setOrganization(state, updatedOrganization)
  },
  'append commuters' (state, action) {
    const {commuters, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', commuters)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete group' (state, action) {
    const {organizationId, groupId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'groups', groupId)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Site Stuff
   */

  'add site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'sites', [site])
    return setOrganization(state, updatedOrganization)
  },
  'delete site' (state, action) {
    const {organizationId, siteId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'sites', siteId)
    return setOrganization(state, updatedOrganization)
  },
  'update site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = setEntity(affectedOrganization, 'sites', site)
    return setOrganization(state, updatedOrganization)
  }
}

// Utilities

/**
 * Helper to delete an item from a entityById object
 * @param  {String} id  The ID of the entity to delete
 * @return {Function}   A function that deletes the entity from
 *  the object and then returns that object
 */
const deleteFromMap = (id) => (map) => {
  delete map[id]
  return map
}

export const initialState = {
  organizations: [],
  organizationsById: {}
}

/**
 * Add entities to a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Array} entities          the entities to add
 * @return                          The updated parent
 */
const addEntities = (parent, entitiesName, entities) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesName]: { $push: entities }
  })
  return update(updatedParent, {
    [entitiesByIdName]: { $set: arrayToObj(updatedParent[entitiesName]) }
  })
}

/**
 * Delete an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Object} entityId         ID of the entity to delete
 * @return                          The updated parent
 */
const deleteEntity = (parent, entitiesName, entityId) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: { $apply: deleteFromMap(entityId) }
  })
  return update(updatedParent, {
    [entitiesName]: { $set: Object.values(updatedParent[entitiesByIdName]) }
  })
}

/**
 * Helper to update an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys* @param {Object} updatedEntity    entity to update
 * @return                          The updated parent
 */
const setEntity = (parent, entitiesName, updatedEntity) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: {
      [updatedEntity.id]: {
        $set: updatedEntity
      }
    }
  })
  return update(updatedParent, {
    [entitiesName]: {
      $set: Object.values(updatedParent[entitiesByIdName])
    }
  })
}

const setGroup = (organization, group) => {
  return setEntity(organization, 'groups', group)
}

const setOrganization = (state, organization) => {
  return setEntity(state, 'organizations', organization)
}
