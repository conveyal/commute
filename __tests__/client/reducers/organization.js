/* globals describe, expect, it */

import {handleActions} from 'redux-actions'

import {
  mockCommuter,
  mockSite,
  mockStores,
  mockTrip,
  simpleOrganization
} from '../../test-utils/mock-data'
import * as organization from '../../../client/reducers/organization'

describe('client > reducers > organization', () => {
  const reducer = handleActions(organization.reducers)

  // Default State Test
  it('should handle default state', () => {
    const result = reducer(organization.initialState, { type: 'blah', payload: {} })
    expect(result.organizations.length).toBe(0)
    expect(result).toMatchSnapshot()
  })

  // Specific Handler Tests
  // Organization Tests
  it('should handle add organization', () => {
    const action = { type: 'add organization', payload: simpleOrganization }
    const result = reducer(organization.initialState, action)
    expect(result.organizations.length).toBe(1)
    expect(result.organizations[0]).toMatchSnapshot()
    expect(result.organizations[0]).toEqual(result.organizationsById['1'])
  })

  it('should handle set organizations', () => {
    const action = { type: 'set organizations', payload: [simpleOrganization] }
    const result = reducer(organization.initialState, action)
    expect(result.organizations.length).toBe(1)
    expect(result.organizations[0]).toMatchSnapshot()
    expect(result.organizations[0]).toEqual(result.organizationsById['1'])
  })

  /**
   * Analysis Tests
   */

  it('should handle add analysis', () => {
    const newAnalysis = {
      id: '2',
      groupId: '1',
      lastRunDateTime: 1477777777,
      name: 'Another Analysis',
      siteId: '1'
    }
    const action = { type: 'add analysis', payload: { organizationId: '2', analysis: newAnalysis } }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedOrganization = result.organizationsById['2']
    expect(affectedOrganization.analyses.length).toBe(2)
    expect(affectedOrganization.analyses[1]).toMatchSnapshot()
    expect(affectedOrganization.analyses[1]).toEqual(affectedOrganization.analysesById['2'])
  })

  it('should handle delete analysis', () => {
    const action = { type: 'delete analysis', payload: { organizationId: '2', analysisId: '1' } }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedOrganization = result.organizationsById['2']
    expect(affectedOrganization.analyses.length).toBe(0)
    expect(affectedOrganization.analysesById).toEqual({})
  })

  it('should handle receive mock calculated trips', () => {
    const calculatedTrip = Object.assign({}, mockTrip, {
      mostLikely: {
        cost: 4.56,
        distance: 22,
        time: 2321,
        mode: 'transit',
        polygon: 'encoded'
      }
    })
    const action = {
      type: 'receive mock calculated trips',
      payload: {
        organizationId: '2',
        analysisId: '1',
        trips: [calculatedTrip]
      }
    }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedAnalysis = result.organizationsById['2'].analysesById['1']
    expect(affectedAnalysis.trips.length).toBe(1)
    expect(affectedAnalysis.tripVals.bike.cost.length).toBe(1)
    expect(affectedAnalysis).toMatchSnapshot()
  })

  /**
   * Commuter Stuff
   */

  const commuterSal = {
    address: '9876 ABC Ct',
    email: 'sal@a.mander',
    id: '2',
    lat: 38.915,
    lng: -76.971,
    name: 'Sal A. Mander'
  }

  it('should handle add commuter', () => {
    const action = {
      type: 'add commuter',
      payload: {
        commuter: commuterSal,
        organizationId: '2',
        groupId: '1'
      }
    }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedGroup = result.organizationsById['2'].groupsById['1']
    expect(affectedGroup.commuters.length).toBe(2)
    expect(affectedGroup.commuters[1]).toMatchSnapshot()
    expect(affectedGroup.commuters[1]).toEqual(affectedGroup.commutersById['2'])
  })

  it('should handle delete commuter', () => {
    const action = {
      type: 'delete commuter',
      payload: { groupId: '1', commuterId: '1', organizationId: '2' }
    }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedGroup = result.organizationsById['2'].groupsById['1']
    expect(affectedGroup.commuters.length).toBe(0)
    expect(affectedGroup.commutersById).toEqual({})
  })

  it('should handle update commuter', () => {
    const action = {
      type: 'update commuter',
      payload: {
        commuter: Object.assign({}, commuterSal, { id: '1' }),
        organizationId: '2',
        groupId: '1'
      }
    }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedGroup = result.organizationsById['2'].groupsById['1']
    expect(affectedGroup.commuters.length).toBe(1)
    expect(affectedGroup.commuters[0]).toMatchSnapshot()
    expect(affectedGroup.commuters[0]).toEqual(affectedGroup.commutersById['1'])
  })

  /**
   * Group Stuff
   */

  it('should handle add group', () => {
    const newGroup = {
      allAddressesGeocoded: true,
      id: '1',
      name: 'Mock Group',
      commuters: [mockCommuter]
    }
    const action = { type: 'add group', payload: { organizationId: '1', group: newGroup } }
    const result = reducer(mockStores.oneSimpleOrganization.organization, action)
    const affectedOrganization = result.organizationsById['1']
    expect(affectedOrganization.groups.length).toBe(1)
    expect(affectedOrganization.groups[0]).toMatchSnapshot()
    expect(affectedOrganization.groups[0]).toEqual(affectedOrganization.groupsById['1'])
  })

  it('should handle append commuters', () => {
    const action = {
      type: 'append commuters',
      payload: {
        commuters: [commuterSal],
        groupId: '1',
        organizationId: '2'
      }
    }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedGroup = result.organizationsById['2'].groupsById['1']
    expect(affectedGroup.commuters.length).toBe(2)
    expect(affectedGroup.commuters[1]).toMatchSnapshot()
    expect(affectedGroup.commuters[1]).toEqual(affectedGroup.commutersById['2'])
  })

  it('should handle delete group', () => {
    const action = { type: 'delete group', payload: { organizationId: '2', groupId: '1' } }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedOrganization = result.organizationsById['2']
    expect(affectedOrganization.groups.length).toBe(0)
    expect(affectedOrganization.groupsById).toEqual({})
  })

  /**
   * Site Stuff
   */

  it('should handle add site', () => {
    const action = { type: 'add site', payload: { organizationId: '1', site: mockSite } }
    const result = reducer(mockStores.oneSimpleOrganization.organization, action)
    const affectedOrganization = result.organizationsById['1']
    expect(affectedOrganization.sites.length).toBe(1)
    expect(affectedOrganization.sites[0]).toMatchSnapshot()
    expect(affectedOrganization.sites[0]).toEqual(affectedOrganization.sitesById['1'])
  })

  it('should handle delete site', () => {
    const action = { type: 'delete site', payload: { organizationId: '2', siteId: '1' } }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedOrganization = result.organizationsById['2']
    expect(affectedOrganization.sites.length).toBe(0)
    expect(affectedOrganization.sitesById).toEqual({})
  })

  it('should handle update site', () => {
    const siteUpdate = Object.assign({}, mockSite, { name: 'Delta Corp', address: '456 DEF Rd' })
    const action = { type: 'update site', payload: { organizationId: '2', site: siteUpdate } }
    const result = reducer(mockStores.complexOrganization.organization, action)
    const affectedOrganization = result.organizationsById['2']
    expect(affectedOrganization.sites[0]).toMatchSnapshot()
    expect(affectedOrganization.sites[0]).toEqual(affectedOrganization.sitesById['1'])
  })
})
