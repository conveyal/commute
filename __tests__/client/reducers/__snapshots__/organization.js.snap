exports[`client > reducers > organization should handle add analysis 1`] = `
Object {
  "groupId": "1",
  "id": "2",
  "lastRunDateTime": 1477777777,
  "name": "Another Analysis",
  "siteId": "1",
}
`;

exports[`client > reducers > organization should handle add commuter 1`] = `
Object {
  "address": "9876 ABC Ct",
  "email": "sal@a.mander",
  "id": "2",
  "lat": 38.915,
  "lng": -76.971,
  "name": "Sal A. Mander",
}
`;

exports[`client > reducers > organization should handle add group 1`] = `
Object {
  "allAddressesGeocoded": true,
  "commuters": Array [
    Object {
      "address": "4321 XYZ Boulevard",
      "email": "luke@warm.cold",
      "id": "1",
      "lat": 38.916089,
      "lng": -76.970221,
      "name": "Luke Warm",
    },
  ],
  "commutersById": Object {
    "1": Object {
      "address": "4321 XYZ Boulevard",
      "email": "luke@warm.cold",
      "id": "1",
      "lat": 38.916089,
      "lng": -76.970221,
      "name": "Luke Warm",
    },
  },
  "id": "1",
  "name": "Mock Group",
}
`;

exports[`client > reducers > organization should handle add organization 1`] = `
Object {
  "analyses": Array [],
  "analysesById": Object {},
  "contact": "someone",
  "email": "abc@def.ghi",
  "groups": Array [],
  "groupsById": Object {},
  "id": "1",
  "logo_url": "https://placekitten.com/80/80",
  "main_url": "https://placekitten.com/",
  "name": "Mock Organization",
  "sites": Array [],
  "sitesById": Object {},
}
`;

exports[`client > reducers > organization should handle add site 1`] = `
Object {
  "address": "123 ABC St",
  "id": "1",
  "lat": 38.8886,
  "lng": -77.043,
  "name": "Acme Corp",
  "radius": 1,
}
`;

exports[`client > reducers > organization should handle append commuters 1`] = `
Object {
  "address": "9876 ABC Ct",
  "email": "sal@a.mander",
  "id": "2",
  "lat": 38.915,
  "lng": -76.971,
  "name": "Sal A. Mander",
}
`;

exports[`client > reducers > organization should handle default state 1`] = `
Object {
  "organizations": Array [],
  "organizationsById": Object {},
}
`;

exports[`client > reducers > organization should handle receive mock calculated trips 1`] = `
Object {
  "groupId": "1",
  "id": "1",
  "lastRunDateTime": 1477697490,
  "name": "An Analysis",
  "siteId": "1",
  "summary": Object {
    "avgDistance": 22,
    "avgTravelTime": 2321,
    "savingsPerTrip": 2.22,
    "savingsPerTripYear": 577.2,
    "savingsTotalPerDay": 2.22,
    "savingsTotalPerYear": 577.2,
  },
  "tripVals": Object {
    "bike": Object {
      "cost": Array [
        0.12,
      ],
      "distance": Array [
        23,
      ],
      "time": Array [
        2345,
      ],
    },
    "car": Object {
      "cost": Array [
        6.78,
      ],
      "distance": Array [
        18,
      ],
      "time": Array [
        1234,
      ],
    },
    "mostLikely": Object {
      "cost": Array [
        4.56,
      ],
      "distance": Array [
        22,
      ],
      "time": Array [
        2321,
      ],
    },
    "transit": Object {
      "cost": Array [
        3.45,
      ],
      "distance": Array [
        30,
      ],
      "time": Array [
        3456,
      ],
    },
    "walk": Object {
      "cost": Array [
        0,
      ],
      "distance": Array [
        19,
      ],
      "time": Array [
        6789,
      ],
    },
  },
  "trips": Array [
    Object {
      "bike": Object {
        "cost": 0.12,
        "distance": 23,
        "polygon": "encoded",
        "time": 2345,
      },
      "car": Object {
        "cost": 6.78,
        "distance": 18,
        "polygon": "encoded",
        "time": 1234,
      },
      "commuterId": "1",
      "mostLikely": Object {
        "cost": 4.56,
        "distance": 22,
        "mode": "transit",
        "polygon": "encoded",
        "time": 2321,
      },
      "transit": Object {
        "cost": 3.45,
        "distance": 30,
        "polygon": "encoded",
        "time": 3456,
      },
      "walk": Object {
        "cost": 0,
        "distance": 19,
        "polygon": "encoded",
        "time": 6789,
      },
    },
  ],
}
`;

exports[`client > reducers > organization should handle set organizations 1`] = `
Object {
  "analyses": Array [],
  "analysesById": Object {},
  "contact": "someone",
  "email": "abc@def.ghi",
  "groups": Array [],
  "groupsById": Object {},
  "id": "1",
  "logo_url": "https://placekitten.com/80/80",
  "main_url": "https://placekitten.com/",
  "name": "Mock Organization",
  "sites": Array [],
  "sitesById": Object {},
}
`;

exports[`client > reducers > organization should handle update commuter 1`] = `
Object {
  "address": "9876 ABC Ct",
  "email": "sal@a.mander",
  "id": "1",
  "lat": 38.915,
  "lng": -76.971,
  "name": "Sal A. Mander",
}
`;

exports[`client > reducers > organization should handle update site 1`] = `
Object {
  "address": "456 DEF Rd",
  "id": "1",
  "lat": 38.8886,
  "lng": -77.043,
  "name": "Delta Corp",
  "radius": 1,
}
`;
